File: UITableViewThreeEventsDemo/Demo-objc/Demo-objc/TagViewController.m
//
//  TagViewController.m
//  Demo-20160419
//
//  Created by crw on 4/19/16.
//  Copyright © 2016 crw. All rights reserved.
//

#import "TagViewController.h"
#import "Cell.h"

@interface TagViewController()

@property (nonatomic, strong) NSMutableArray     *dataArray;

@end

@implementation TagViewController

#pragma mark - life cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.dataArray = [Model getData];
    
    [self.tableView reloadData];
}

#pragma mark - UITableViewDelegate
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
  
    static NSString *identifier = @"Cell";
    
    Cell *cell = [tableView dequeueReusableCellWithIdentifier:identifier forIndexPath:indexPath];
    
    cell.model = self.dataArray[indexPath.row];
    
    [cell.delBtn addTarget:self action:@selector(delItem:) forControlEvents:UIControlEventTouchUpInside];
    cell.delBtn.tag = indexPath.row;
    
    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.dataArray.count;
}

#pragma mark - Events
//可以正确进行删除操作，没有删除动画
- (void)delItem:(UIButton *)btn{
    
    NSInteger index = btn.tag;
    [self.dataArray removeObjectAtIndex:index];
    [self.tableView reloadData];
}

//第一次可以正常，后续会出错，有crash
//- (void)delItem:(UIButton *)btn{
//
//    NSInteger index = btn.tag;
//    [self.dataArray removeObjectAtIndex:index];
//    [self.tableView deleteRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]] withRowAnimation:UITableViewRowAnimationMiddle];
//}

//第一次可以正常，后续会出错，不会crash
//- (void)delItem:(UIButton *)btn{
//    
//    NSInteger index = btn.tag;
//    if (index < self.dataArray.count) {
//        [self.dataArray removeObjectAtIndex:index];
//        [self.tableView deleteRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]] withRowAnimation:UITableViewRowAnimationMiddle];
//    }else{
//        NSLog(@" ");
//        NSLog(@"－－－－越界－－－－");
//        NSLog(@"当前index -> %ld",(long)index);
//        NSLog(@"array count -> %lu",(unsigned long)self.dataArray.count);
//    }
//}
@end

File: UITableViewThreeEventsDemo/Demo-swift/Demo-swift/TagViewController.swift
//
//  TagViewController.swift
//  Demo-20160418
//
//  Created by crw on 4/20/16.
//  Copyright © 2016 crw. All rights reserved.
//

import UIKit

let ID = "Cell"

class TagViewController: UITableViewController {
    
    var dataArray = [Model]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        dataArray = Model.getData()
        
        tableView.reloadData()
    }
    
}

extension TagViewController
{

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataArray.count
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(ID, forIndexPath: indexPath) as! Cell
        cell.model = dataArray[indexPath.row]
        cell.delBtn.addTarget(self, action:#selector(deleteItem(_:)), forControlEvents: UIControlEvents.TouchUpInside)
        cell.delBtn.tag = indexPath.row
        return cell
    }
    
    //没有删除动画,正常运行
    func deleteItem(btn : UIButton) {
        
        let index = btn.tag;
        
        dataArray.removeAtIndex(index)
        tableView.reloadData()
    }
    
//    //第一次可以正常，后续会出错，有删除动画,不会crash
//    func deleteItem(btn : UIButton) {
//        
//        let index = btn.tag;
//        
//        if (index < self.dataArray.count) {
//            dataArray.removeAtIndex(index)
//            tableView.deleteRowsAtIndexPaths([NSIndexPath.init(forRow: index, inSection: 0)], withRowAnimation: UITableViewRowAnimation.Fade)
//        }else{
//            print(" ");
//            print("－－－－越界－－－－")
//            print("当前index -> \(index)")
//            print("array count -> \(dataArray.count)")
//        }
//    }
    
//    //第一次可以正常，后续会出错，有删除动画,会crash
//    func deleteItem(btn : UIButton) {
//        
//        let index = btn.tag;
//        
//        dataArray.removeAtIndex(index)
//        tableView.deleteRowsAtIndexPaths([NSIndexPath.init(forRow: index, inSection: 0)], withRowAnimation: UITableViewRowAnimation.Fade)
//    }
    
}
